_drafts/unittests-cpputest.md

```diff
@@ -0,0 +1,1251 @@
+as the framework has the minimum features listed
+[above](#testing-without-a-framework), it is as good as any.
+
+{:.no_toc}
+
+### Initial Setup
+
+We first need to install a pre-compiled version of CppUTest so we can easily run
+tests without needing to compile the binary ourselves from source before every
+test run. The easiest way to do this is to use your system's package manager.
+
+On macOS, CppUTest can be installed using `brew`:
+
+```
+$ brew install cpputest
+```
+
+On Ubuntu, it can be installed using `apt`:
+
+```
+$ sudo apt-get install cpputest
+```
+
+{:.no_toc}
+
+### Project CppUTest Harness
+
+Since it is a decent amount of boilerplate due to the CppUTest harness setup
+required, this example will start with a clone of the example repository and
+then go over the components within it, briefly covering the CppUTest Makefiles.
+
+Feel free to use the code in any way you like, and even copy it into your
+project. It should build quite easily once some paths are patched up.
+
+```
+$ git clone https://github.com/memfault/interrupt.git
+$ cd examples/cpputest-unittests/minimal/tests
+
+# macOS
+$ make
+
+# Ubuntu
+make CPPUTEST_HOME=/usr TARGET_PLATFORM=x86_64-linux-gnu
+
+compiling test_my_sum.cpp
+compiling AllTests.cpp
+...
+Linking build/sum/sum_tests
+Running build/sum/sum_tests
+.
+OK (1 tests, 1 ran, 1 checks, 0 ignored, 0 filtered out, 1 ms)
+```
+
+To set CPPUTEST_HOME and TARGET_PLATFORM for your platform, edit the first few
+lines of `MarkefileWorkerOverrides.mk`.
+
+```
+CPPUTEST_HOME ?= /usr/local/Cellar/cpputest/3.8
+TARGET_PLATFORM ?=
+```
+
+## Tips & Tricks
+
+{:.no_toc}
+
+### Debugging a Unit Test
+
+Most unit test frameworks will generate separate binaries for each `.cpp` unit
+test file written so that you can load them in a debugger (lldb or gdb).
+
+```
+$ lldb build/kv_store/kv_store_tests
+(lldb) target create "build/kv_store/kv_store_tests"
+Current executable set to 'build/kv_store/kv_store_tests' (x86_64).
+(lldb) run
+Process 96031 launched 'build/kv_store/kv_store_tests' (x86_64)
+.
+OK (1 tests, 1 ran, 5 checks, 0 ignored, 0 filtered out, 7 ms)
+
+Process 96031 exited with status = 0 (0x00000000)
+```
+
+{:.no_toc}
+
+### Code Coverage
+
+One of the wonderful parts about unit testing is that you can generate a code
+coverage report. This shows which paths were covered in a given set of unit
+tests, so you can be sure that the piece of code was tested in some capacity.
+Note that code coverage doesn't measure the different behaviors a code path
+**could take**, but only that a particular code path **was taken**.
+
+![](img/cpputest-unittests/code_coverage.png)
+
+Above is a an example coverage report taken from the Memfault Public SDK[^5].
+
+{:.no_toc}
+
+### Address Sanitizing
+
+To raise use-after-free and buffer overflow errors in unit tests, use the
+compiler option `-fsanitize=address` when compiling unit tests.
+
+[AddressSanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer)
```

Comment:
This link is kind of weird. Do you want to just write "You can find out more in the AddressSanitizer documentation" and add a footnote?